### DISCUSSION TITLE
Code Freeze Policy?

### DISCUSSION THREAD

**Discussion by DanVanAtta** [triplea-game/triplea#12150] [General]
@Cernelius  said in #12140

> Maybe off-topic, and I'm not a developer, but, if (I hope) the release is getting close, wouldn't it be better to forbid any additional features (like https://github.com/triplea-game/triplea/pull/12084) until TripleA is released?

> I mean, I was thinking of something like a 3 months freeze on new features whenever it is estimated that a release should happen within 3 months, not blocking new features for a long time. I don't actually know if TripleA is expected to be released by February or whenever though.

**Comment by DanVanAtta** [comment]
@Cernelius one saying about code freezes,  "after code freeze comes bug thaw."

This forum thread went into detail on the changing deployment & release strategy:
https://forums.triplea-game.org/topic/3304/triplea-project-update-2-6-status-changes-to-release-strategy/37

**Reply by DanVanAtta** [reply]
Hmm, the important first page of that link is not rendering..

The post explains:
- too long went between releases, we have not great way to solve the problem from happening again
- deployment & release can be separated, when a feature is made available can happen after it is merged to code, see: https://martinfowler.com/articles/feature-toggles.html (Check sections like: the text part of the intro, "Getting ready to release
" & "Categories of toggles")
To resolve those, we can instead 'release' after every merge to master, always. This will help force us keep backward compatibility and forcibly would solve the problem of too long between releases.

**Comment by Cernelius** [comment]
Now it works.

I remember I saw that, but then I completely forgot.

I guess my suggestion can be reworded as a one-time thing for this release only then (not to add features for now if the release is a few months ahead).

I'll have to re-read it and don't remember if I replied, but I think that maybe I tend to agree because (as a map-maker) remaining stuck for years between significantly different released and to-be-released versions is terrible.

**Comment by asvitkine** [comment]
So here are my thoughts:

I am of the strong opinion that we should try to get 2.6 released asap (fixing whatever blockers remain).
Assuming that's what we're doing, then I agree we should avoid any large risky changes at this point (e.g. things like the combat rewrite of 2.6 - as those can destabilize things significantly).

However, I think small improvements can still be OK - and I wouldn't necessary block things like the pending purchase save game support that was just added, since the change ended up being pretty small - and yes it introduced an issue, but it was a pretty benign / easy fix.

Of course, if we're really at the point where we would be a week or days away from a release, I can see the argument for blocking those too.

All in all, I'd like us to reach a new "stable milestone" with the release of 2.6 so that we can clean out the bug database of old pre-2.6 issues and use that as the new baseline for determining regressions, etc.

Separately, I'm not actually convinced of the "every change is a release and everything must be backwards compatible" proposal by Dan, but that's for after the 2.6 release and we can discuss more later. I'd rather we go with some predictable release schedule (e.g. quarterly) from HEAD and have the flexibility to break backwards compatibility sometimes. Especially since the engine still uses some fragile formats like Java serialized objects, that would really lock us into the current class model if we have to keep backwards compatibility indefinitely. If we get away from those and just have actual data formats, then I can see backwards compatibility being more reasonable to maintain, though. (Having said that, I do think breaking backwards compatibility shouldn't be done lightly, I just disagree that we should never break it.)

Anyway, those are my 2c.

**Reply by DanVanAtta** [reply]
2.6 was intended to be a quarterly release. It's part of the issue, reliance on people to be available & execute a SOP is a challenge. I do think a set release cadence would have a lot of benefits. At the same time, that was the intent of 2.6 and it did not work out well. My impression is that release issues tend to be found pretty late,notably when wanting to do that next release. Typically, that will be after some developer has spent a lot of time working on something, then spent 3x longer on it because it's TripleA, and now they simply don't have time. Or alternatively, fixing & dealing with the release takes up all the time for developers and they don't get to do much more than maintainer activities. 

I personally don't know how to solve that problem here.  The every-release-goes-to-prod situation is a solution to some of that, and should help reduce latency between problem introduction, detection and fix. I suspect there could be additional benefits, such as the attractiveness of being able to see changes live very quickly and in a predictable manner.

On a slightly different note, as an important clarification, I did not mean to convey that we would never break backward compatibility. At the same time, there are a lot of techniques that we can use to avoid the problem, EG: conversion tools, multiple lobbies, feature flags.

Compatibility is a a nuanced conversation at the same time. There are multiple dimensions of compatibility, eg:  client vs another client, client vs lobby, client vs save-game, client vs map-version.  These different dimensions have different approaches to them..

I do think there will be some problems at the same time. There will be something merged that does break compat, and then we find out a month later and it's then not-fun to debug. At the same time, we have the same problem now and we tend to find out even later.

**Comment by Cernelius** [comment]
Yeah, I think it has been amply proved that quarterly or whatever release schedules have failed again and again, and I don't even remember how many years the 2.6 release is overdue...

In a project in which anyone can walk away any time, you probably have to release any time as the pain of having wildly unpredictable release times is going to be greater than the still likely egregious and chronic unreliability of a constant release model.

Any opinions on whether 2.6 would warrant being numbered 3.0 upon release (also for entering a new release model era)?

**Reply by DanVanAtta** [reply]
I would overall agree, yet in fairness the project never formalized a release schedule.
